staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -315,23 +320,96 @@ func strategicPatchObject(
 
 type applyPatcher struct {
 	*patcher
+
+	model proto.Schema
 }
 
-func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
-	versionedObjToUpdate, err := p.creater.New(p.kind)
+// This is not going to stay, but we don't have a way to pass the
+// workflow id so far, so let's always use this one...
+const workflowId = "default"
+
+func (p *applyPatcher) LiveObject(obj runtime.Object) (map[string]interface{}, error) {
+	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.kind.GroupVersion())
 	if err != nil {
 		return nil, err
 	}
-	if err := runtime.DecodeInto(p.codec, p.patchBytes, versionedObjToUpdate); err != nil {
-		return nil, err
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
+}
+
+func (p *applyPatcher) LastAppliedObject(obj runtime.Object, workflow string) (map[string]interface{}, error) {
+	acsor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
 	}
-	// For now, instead of performing a structured merge we will just attempt a PUT
-	// So we just convert the object back to unversioned (aka internal version).
-	unversionedObjToUpdate, err := p.toUnversioned(versionedObjToUpdate)
+	last := make(map[string]interface{})
+	if acsor.GetLastApplied()[workflow] != "" {
+		if err := json.Unmarshal([]byte(acsor.GetLastApplied()[workflow]), &last); err != nil {
+			return nil, fmt.Errorf("couldn't unmarshal last applied field: %v", err)
+		}
+	}
+	return last, nil
+}
+
+func (p *applyPatcher) PatchObject() (map[string]interface{}, error) {
+	patch := make(map[string]interface{})
+	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
+	}
+	return patch, nil
+}
+
+func (p *applyPatcher) OutputObject(result interface{}) (runtime.Object, error) {
```

Comment:
unstructuredVersionedToTypedUnversioned?

Reply:
I kind of liked the fact that we have one method to get each type of object (focusing on the what rather than the how).

Reply:
Oh, you meant this name to be a noun, too?

The other three are just getters, but this one transforms an input parameter, so I don't think a noun name will be clear. (just a hot take)